src/modROS.lua

```diff
@@ -86,21 +80,24 @@ function ModROS:loadMap()
     self.path = ModROS.MOD_DIR .. "ROS_messages"
     self._conx = WriteOnlyFileConnection.new(self.path)
 
-    -- initialise publishers
+    -- initialise no-namespace-specific publishers
     self._pub_tf = Publisher.new(self._conx, "tf", tf2_msgs_TFMessage)
     self._pub_clock = Publisher.new(self._conx, "clock", rosgraph_msgs_Clock)
-    self._pub_odom = Publisher.new(self._conx, "odom", nav_msgs_Odometry)
-    self._pub_scan = Publisher.new(self._conx, "scan", sensor_msgs_LaserScan)
-    self._pub_imu = Publisher.new(self._conx, "imu", sensor_msgs_Imu)
 
     print("modROS (" .. ModROS.MOD_VERSION .. ") loaded")
 end
 
 function ModROS.installSpecializations(vehicleTypeManager, specializationManager, modDirectory, modName)
     specializationManager:addSpecialization("rosVehicle", "RosVehicle", Utils.getFilename("src/vehicles/RosVehicle.lua", modDirectory), nil) -- Nil is important here
 
-    for typeName, _ in pairs(vehicleTypeManager:getVehicleTypes()) do
-        vehicleTypeManager:addSpecialization(typeName, modName .. ".rosVehicle")
+    for typeName, typeEntry in pairs(vehicleTypeManager:getVehicleTypes()) do
+        -- only if the prerequisites of specialization are fulfilled, specializations <rosVehicle> will be installed
+        -- in RosVehicle.lua, spec <drivable> is set as one of prerequisites for now
+        -- hence, if there exists no this condition, an error will occur: Not all prerequisites of specialization modROS.rosVehicle are fulfilled
+
+        if SpecializationUtil.hasSpecialization(Drivable, typeEntry.specializations) then
+            vehicleTypeManager:addSpecialization(typeName, modName .. ".rosVehicle")
```

Comment:
So just for my understanding: what's the relation between `installSpecializations(..)` and `RosVehicle.prerequisitesPresent(..)`?

Having to check two times for `Drivable` seems strange to me. I sort-of expected `RosVehicle.prerequisitesPresent(..)` to return `false` for anything without `Drivable`, and that fact would make FS conclude it should not continue trying to install the spec for that entity.

But that's not how it works I guess from reading this code?

Reply:
Ping?

Reply:
AFAIK, `installSpecializations(..)` part is to actually get `rosVehicle` spec on a vehicle type. And RosVehicle.prerequisitesPresent() is to verify if a vehicle has a specific specialization.

if the below condition is not added, the `rosVehicle` will be added to all vehicle types regardless of whether it has drivable spec as prerequisite or not, and an error will occur when running `RosVehicle.prerequisitesPresent()`

`if SpecializationUtil.hasSpecialization(Drivable, typeEntry.specializations) then`